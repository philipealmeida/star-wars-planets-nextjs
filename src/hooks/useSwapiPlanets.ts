import useSWR from 'swr';
import { fetchPlanets } from '@/lib/api';
import type { PlanetsResponse } from '@/lib/types';

export function useSwapiPlanets(page: number, search?: string) {
  const { data, error, isLoading } = useSWR<PlanetsResponse>(
    ['planets', page, search],
    () => fetchPlanets(page, search)
  );

  return {
    planets: data?.results,
    count: data?.count,
    isLoading,
    error,
  };
}