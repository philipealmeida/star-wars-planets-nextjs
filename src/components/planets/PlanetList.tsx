'use client';
import { useEffect, useState } from 'react';
import Pagination from './Pagination';
import PlanetCard from './PlanetCard';
import PlanetSearch from './PlanetSearch';
import { useRouter } from 'next/navigation';
import { useDebounce } from '@/hooks/useDebounce';
import { useSearchParams } from 'next/navigation';
import { useSwapiPlanets } from '@/hooks/useSwapiPlanets';
import styles from '@/styles/components/planets/PlanetList.module.css';

const SEARCH_DEBOUNCE_DELAY = 500;

export default function PlanetList() {
	const router = useRouter();
	const searchParams = useSearchParams();

	const search = searchParams.get('search') || '';
	const page = Number(searchParams.get('page')) || 1;
	const [lastSearchedTerm, setLastSearchedTerm] = useState(search);
	const [isSearching, setIsSearching] = useState(false);
	const debouncedSearch = useDebounce(search, SEARCH_DEBOUNCE_DELAY);

	const { planets, count, isLoading, error } = useSwapiPlanets(page, debouncedSearch);

	const handleSearch = (value: string) => {
		if (value !== lastSearchedTerm) {
			setLastSearchedTerm(value);
			const params = new URLSearchParams();

			if (value) {
				params.set('search', value);
			}

			params.set('page', '1');
			router.push(`/planets?${params.toString()}`);
		}
	};

	const handlePageChange = (newPage: number) => {
		const params = new URLSearchParams(searchParams);
		params.set('page', newPage.toString());
		router.push(`/planets?${params.toString()}`);
	};

	useEffect(() => {
		if (search !== debouncedSearch) {
			setIsSearching(true);
		} else {
			setIsSearching(false);
		}
	}, [search, debouncedSearch]);

	if (error) return <div className={styles.error}>Failed to load planets</div>;

	return (
		<section className={styles.container}>
			<PlanetSearch value={search} onChange={handleSearch} />
			{(isLoading || isSearching) ? (

				<div className={styles.loading}>Loading...</div>
			) : (
				<>
					<div className={`${styles.content} ${!planets?.length ? styles.empty : ''}`}>
						{planets?.length ? (
							planets.map((planet) => (
								<PlanetCard key={planet.url} planet={planet} />
							))
							) : (
							<div className={styles.empty}>No planets found</div>
						)}
					</div>
					<Pagination
						currentPage={page}
						totalPages={Math.ceil((count || 0) / 10)}
						onPageChange={handlePageChange}
					/>
				</>
			)}
		</section>
	);
}