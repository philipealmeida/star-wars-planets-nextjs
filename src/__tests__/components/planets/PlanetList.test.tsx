import PlanetList from '@/components/planets/PlanetList';
import { useSwapiPlanets } from '@/hooks/useSwapiPlanets';
import { render, screen } from '@testing-library/react';

jest.mock('next/navigation', () => ({
  useRouter: () => ({
    push: jest.fn(),
  }),
  useSearchParams: () => ({
    get: jest.fn().mockImplementation((param) => {
      if (param === 'page') return '1';
      if (param === 'search') return '';
      return null;
    }),
  }),
}));

jest.mock('@/hooks/useSwapiPlanets');
jest.mock('@/hooks/useDebounce', () => ({
  useDebounce: (value: string) => value,
}));

const mockPlanets = [
  {
    name: 'Tatooine',
    climate: 'Arid',
    terrain: 'Desert',
    population: '200000',
    url: 'https://swapi.dev/api/planets/1/',
    diameter: '10465',
    gravity: '1 standard',
    surface_water: '1',
  },
];

const renderComponent = () => {
  return render(<PlanetList />);
};

describe('PlanetList', () => {
  beforeEach(() => {
    (useSwapiPlanets as jest.Mock).mockReturnValue({
      planets: mockPlanets,
      count: 1,
      isLoading: false,
      error: null,
    });
  });

  it('matches snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('displays loading state', () => {
    (useSwapiPlanets as jest.Mock).mockReturnValue({
      planets: [],
      count: 0,
      isLoading: true,
      error: null,
    });
    
    renderComponent();
    expect(screen.getByText('Loading...')).toBeInTheDocument();
  });

  it('displays error state', () => {
    (useSwapiPlanets as jest.Mock).mockReturnValue({
      planets: [],
      count: 0,
      isLoading: false,
      error: new Error('Failed to fetch'),
    });
    
    renderComponent();
    expect(screen.getByText('Failed to load planets')).toBeInTheDocument();
  });

  it('renders planet cards when data is loaded', () => {
    renderComponent();
    expect(screen.getByText('Tatooine')).toBeInTheDocument();
  });

  it('displays no planets found message when empty', () => {
    (useSwapiPlanets as jest.Mock).mockReturnValue({
      planets: [],
      count: 0,
      isLoading: false,
      error: null,
    });
    
    renderComponent();
    expect(screen.getByText('No planets found')).toBeInTheDocument();
  });
});