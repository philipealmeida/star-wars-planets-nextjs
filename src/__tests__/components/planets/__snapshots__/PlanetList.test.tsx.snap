// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlanetList matches snapshot 1`] = `
<div>
  <section
    class="container"
  >
    <div
      class="search"
    >
      <input
        class="input"
        placeholder="Search planets..."
        type="text"
        value=""
      />
    </div>
    <div
      class="content "
    >
      <a
        class="card"
        href="/planets/1"
      >
        <h2>
          Tatooine
        </h2>
        <div
          class="details"
        >
          <p>
            Climate: 
            Arid
          </p>
          <p>
            Terrain: 
            Desert
          </p>
          <p>
            Population: 
            200000
          </p>
        </div>
      </a>
    </div>
    <div
      class="pagination"
    >
      <button
        class="button"
        disabled=""
      >
        Previous
      </button>
      <span
        class="info"
      >
        Page 
        1
         of 
        1
      </span>
      <button
        class="button"
        disabled=""
      >
        Next
      </button>
    </div>
  </section>
</div>
`;
